### ISSUE TITLE
`@RegisterArgumentFactory` not honored for handles returned by `GetHandle` methods

### ISSUE THREAD

**Issue by zalimeni** [jdbi/jdbi#830]
It appears that `@RegisterArgumentFactory` does not apply to handles retrieved via `getHandle`/`withHandle` mixed into a `SqlObject` class via the `GetHandle` interface. If so, this seems like at best unintuitive behavior, and at worst dangerous. One particular danger seems to be that arguments used in `WHERE` clauses can be bound incorrectly using a fallback argument factory and silently fail queries (which appears to have happened in my case with a UUID argument).

In `@SqlUpdate`/`@SqlQuery`/`@SqlBatch` methods, a `CustomizingStatementHandler` gets a `SqlStatementCustomizer` to register annotation-provided argument factories to generated `SQLStatement`s. I would expect a DAO that mixes in `GetHandle` to receive handles with the same argument factories registered, in accordance with method- or class-level `@RegisterArgumentFactory` annotations. It seems like this could be done within the `PassThroughHandler`, and `PassThroughTransactionHandler` for `@Transaction` methods. Maybe they could both extend a base passthrough handler class to do this in one place.

Please correct me if I'm missing something that makes any of this inaccurate. If the proposed solution sounds reasonable, I'm happy to open a PR at some point when I have time.

**Comment by qualidafial** [comment]
@mzalimeni This is in v2.x, correct?

**Comment by qualidafial** [comment]
The problem you're describing is actually more general: in v2, annotations are not honored on default methods.

This is already fixed in v3. See e.g. the `TestRegisteredMappersWork.testBeanMapperFactoryDefaultMethod()` unit test.

At this point pretty much all of our focus is on getting v3 ready to ship. However if you're willing to do the work to fix this in v2, I'm definitely willing to provide guidance and code reviews as needed.

**Comment by stevenschlansker** [comment]
Note that v2 assumes Java source level 1.6 which has no concept of default methods at all.  So backporting this feature to v2 may end up to be rather difficult.

**Comment by qualidafial** [comment]
Sometimes I forget that once a Java existed without default methods.

The issue also applies to concrete (non-abstract) class methods

**Comment by mzalimeni** [comment]
Apologies for the slow response time, been struggling to find spare time to loop back to this. I've got v2 checked out and a 1.6 dev environment set up, and I spent a little time this week looking through things. I need to look at it a bit more to grok the generalizable bits that would be applied to handles in a fix. Any suggestions on where to start are welcome if you have them - @qualidafial, thanks for the offer of help, and that sounds great.

I'll update if/when I have something. I'll be off the grid in a few weeks until mid-August, but if you don't mind a slower pace, I may be able to get this out at some point. Regardless, I appreciate the quick replies.
